src/main/java/bio/overture/ego/service/RefreshContextService.java

```diff
@@ -0,0 +1,169 @@
+
+  private RefreshTokenRepository refreshTokenRepository;
+  private TokenService tokenService;
+  /** Configuration */
+  private int durationMs;
+
+  @Autowired
+  public RefreshContextService(
+      @NonNull RefreshTokenRepository refreshTokenRepository,
+      @NonNull TokenService tokenService,
+      @Value("${refreshToken.durationMs:43200000}") int durationMs) {
+    super(RefreshToken.class, refreshTokenRepository);
+    this.refreshTokenRepository = refreshTokenRepository;
+    this.tokenService = tokenService;
+    this.durationMs = durationMs;
+  }
+
+  @SuppressWarnings("unchecked")
+  public RefreshToken get(@NonNull UUID id, boolean fetchUser) {
+    val result =
+        (Optional<RefreshToken>)
+            getRepository()
+                .findOne(new RefreshTokenSpecificationBuilder().fetchUser(fetchUser).buildById(id));
+    checkNotFound(result.isPresent(), "RefreshToken '%s' does not exist", id);
+    return result.get();
+  }
+
+  @Override
+  public RefreshToken getWithRelationships(@NonNull UUID userId) {
+    return get(userId, true);
+  }
+
+  private RefreshToken createTransientToken(UUID jti) {
+    val now = Instant.now();
+    val expiry = now.plus(durationMs, ChronoUnit.MILLIS);
+    return RefreshToken.builder()
+        .jti(jti)
+        .issueDate(Date.from(now))
+        .expiryDate(Date.from(expiry))
+        .build();
+  }
+
+  public Optional<RefreshToken> getRefreshTokenByUser(User user) {
+    return refreshTokenRepository.getByUser(user);
+  }
+
+  private void checkUniqueByUser(User user) {
+    checkUnique(
+        !getRefreshTokenByUser(user).isPresent(),
+        String.format("A refresh token already exists for %s", user.getId()));
+  }
+
+  public void disassociateUserAndDelete(String userToken) {
+    val user = tokenService.getTokenUserInfo(userToken);
+
+    val refreshTokenOpt = this.getRefreshTokenByUser(user);
+    if (refreshTokenOpt.isPresent()) {
+      log.debug("Refresh token exists, deleting...");
+      user.setRefreshToken(null);
+      val existingRefreshToken = refreshTokenOpt.get();
+      refreshTokenRepository.delete(existingRefreshToken);
+    }
+  }
+
+  public RefreshToken createRefreshToken(@NonNull String bearerToken) {
+    val user = tokenService.getTokenUserInfo(bearerToken);
+    this.checkUniqueByUser(user);
+
+    val jti = tokenService.getTokenClaims(bearerToken).getId();
+    val refreshToken = createTransientToken(UUID.fromString(jti));
+    refreshToken.associateWithUser(user);
+    return refreshTokenRepository.save(refreshToken);
+  }
+
+  public RefreshContext createRefreshContext(String refreshTokenId, String bearerToken) {
+    val refreshTokenOpt = refreshTokenRepository.findById(UUID.fromString(refreshTokenId));
+    val tokenClaims = tokenService.getTokenClaims(bearerToken);
+    val user = tokenService.getTokenUserInfo(bearerToken);
+
+    checkNotFound(
+        !isNull(refreshTokenOpt) && refreshTokenOpt.isPresent(),
```

Comment:
You don't need `!isNull(refreshTokenOpt)` in this check, as refreshTokenOpt is an optional type, it will never be null.